Implement local watcher (4/5): Handle progressed and concluded events

manoranjith opened the pull request:
Builds on top of #212, #213, #214

ggwpez commented on watcher/local/watcher.go:
```diff
@@ -0,0 +1,275 @@
+		case *channel.ConcludedEvent:
+			log.Debugf("Received concluded event from chain: %v", e)
+			eventsToClientPubSub.publish(e)
+		default:
```
Could you also log this case, that an event was ignored?

matthiasgeihs replied:
Log.Warn or Error that an unknown event type was observed. This should never happen.

manoranjith replied:
Logged as error. See 5088c4c.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -141,6 +141,22 @@ func Test_Watcher_Working(t *testing.T) {
+			w := newWatcher(t, rs)
+			// Start watching and publish states.
+			statesPub, eventsForClient := startWatchingForLedgerChannel(t, w, makeSignedStateWDummySigs(params, txs[0].State))
+			require.NoError(t, statesPub.Publish(txs[1]))
```
I think the state publishing is not needed here.

manoranjith replied:
Yes, I have removed it. See d20c6b0.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -141,6 +141,22 @@ func Test_Watcher_Working(t *testing.T) {
 			rs.AssertExpectations(t)
 		})
+
+		t.Run("happy/progressed_event", func(t *testing.T) {
```
The tests are very comprehensive (which is good) but could you please add a comment on each subtest what it does? One line would be enough.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -280,6 +297,66 @@ func Test_Watcher_Working(t *testing.T) {
+			// Child: Start watching and publish states.
+			childSignedState := makeSignedStateWDummySigs(childParams, childTxs[0].State)
+			statesPubChild, eventsForClientChild := startWatchingForSubChannel(t, w, childSignedState, parentTxs[0].State.ID)
+			require.NoError(t, statesPubChild.Publish(childTxs[1]))
```
Both state publishing also not needed here AFAIK.

manoranjith replied:
Updated in d20c6b0.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -280,6 +297,66 @@ func Test_Watcher_Working(t *testing.T) {
+			// Parent: Start watching and publish states.
+			parentSignedState := makeSignedStateWDummySigs(parentParams, parentTxs[0].State)
+			statesPubParent, eventsForClientParent := startWatchingForLedgerChannel(t, w, parentSignedState)
+			require.NoError(t, statesPubParent.Publish(parentTxs[1]))
```
not needed

manoranjith replied:
Updated in d20c6b0.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -424,6 +501,36 @@ func makeRegisteredEvent(txs ...channel.Transaction) []channel.AdjudicatorEvent
 	return events
 }
 
+func makeProgressedEvent(txs ...channel.Transaction) []channel.AdjudicatorEvent {
```
`makeProgressedEvents`

manoranjith replied:
Updated in d20c6b0.

ggwpez commented on watcher/local/watcher_test.go:
```diff
@@ -424,6 +501,36 @@ func makeRegisteredEvent(txs ...channel.Transaction) []channel.AdjudicatorEvent
+	return events
+}
+
+func makeConcludedEvent(txs ...channel.Transaction) []channel.AdjudicatorEvent {
```
`makeConcludedEvents`

manoranjith replied:
Updated in d20c6b0.

matthiasgeihs commented on watcher/local/watcher.go:
```diff
@@ -217,6 +217,12 @@ func (w *Watcher) handleEventsFromChain(eventsFromChainSub channel.AdjudicatorSu
 					log.Debug("Registered successfully")
 				}
 			}()
+		case *channel.ProgressedEvent:
```
I was just wondering: Why is the watcher forwarding events to the client at all? Isn't a watcher usually just a consumer?

manoranjith replied:
Events are forwarded to the client as an information. Because client could act based on this information.

1. When a state is registered, it can stop creating newer off-chain states.
2. Optionally, it could also progress or settle the channel.

manoranjith commented:
Squashed the review commits and rebased onto tip of target branch.
Previously the target branch was wrongly set to `dev`, I have changed it to `172-refactor-watcher`.

It is ready to be merged. @matthiasgeihs
